text/0080-datetime-extension.md

```diff
@@ -0,0 +1,367 @@
+  - `"YYYY-MM-DDThh:mm:ssZ"` (UTC)
+  - `"YYYY-MM-DDThh:mm:ss.SSSZ"` (UTC with millisecond precision)
+  - `"YYYY-MM-DDThh:mm:ss(+/-)hhmm"` (With timezone offset in hours and minutes)
+  - `"YYYY-MM-DDThh:mm:ss.SSS(+/-)hhmm"` (With timezone offset in hours and minutes and millisecond precision)
+- `.add(duration)` returns a new `datetime`, offset by duration.
+- `.toDate()` returns a new `datetime`, truncating to the day, such that printing the `datetime` would have `00:00:00` as the time.
+- `.toTime()` returns a new `duration`, removing the days, such that only milliseconds since `.toDate()` are left. This is equivalent to `DT - DT.toDate()`
+
+Values of type `datetime` can be used with comparison operators:
+
+- `DT1 < DT2` returns `true` when `DT1` is before `DT2`
+- `DT1 <= DT2` returns `true` when `DT1` is before or equal to `DT2`
+- `DT1 > DT2` returns `true` when `DT1` is after `DT2`
+- `DT1 >= DT2` returns `true` when `DT1` is after or equal to `DT2`
+- `DT1 == DT2` returns `true` when `DT1` is equal to `DT2`
+
+Values of type `datetime` can also be used with the `-` operator:
+
+- `DT1 - DT2` returns the difference between `DT1` and `DT2` as a `duration`.
+
+Note that the inverse of `-` is `.add(duration)`, not `+`.
+
+The `datetime` type is internally represented as a `long` and contains a Unix Time in milliseconds. This is the number of non-leap seconds that have passed since `1970-01-01T00:00:00Z` in milliseconds. Unix Time days are always 86,400 seconds and handle leap seconds by absorbing them at the start of the day. Due to using Unix Time, and not providing a "current time" function, Cedar avoids the complexities of leap second handling, pushing them to the system and application.
+
+- `duration(long, string)` constructs a duration value. The string argument must be one of `"days", "hours", "minutes", "seconds", "milliseconds"`. Strict validation requires `string` to be a string literal, although evaluation/authorization support any string-typed expression.
+- `.toMilliseconds()` returns a `long` describing the number of milliseconds in this duration. (the value as a long, itself)
+- `.toSeconds()` returns a `long` describing the number of seconds in this duration. (`.toMilliseconds() / 1000`)
+- `.toMinutes()` returns a `long` describing the number of minutes in this duration. (`.toSeconds() / 60`)
+- `.toHours()` returns a `long` describing the number of hours in this duration. (`.toMinutes() / 60`)
+- `.toDays()` returns a `long` describing the number of days in this duration. (`.toHours() / 24`)
+
+Values with type `duration` duration can also be used with comparison operators:
+
+- `DT1 < DT2` returns `true` when `DT1` is before `DT2`
+- `DT1 <= DT2` returns `true` when `DT1` is before or equal to `DT2`
+- `DT1 > DT2` returns `true` when `DT1` is after `DT2`
+- `DT1 >= DT2` returns `true` when `DT1` is after or equal to `DT2`
+- `DT1 == DT2` returns `true` when `DT1` is equal to `DT2`
```

Comment:
This is a cut&paste error from above. Change `DT` to be `dur` and change explanations appropriately, e.g.,

- `dur1 < dur2` evaluates to `true` when `dur1` is shorter than `dur2`

Reply:
Also: Why have `==` here (and above, for DateTimes) but not `!=`? I'd suggest having neither or both. Presumably `==` (and `!=`) continue to mean syntactic (non)equality.

Reply:
All Cedar values currently support `==` and `!=`.  I think it would be very confusing if we didn't support both of those for both duration and datetime.

Reply:
To clarify: `==` and `!=` will work for datetime & duration the same way they do for any other value. They are not being overloaded here -- I'll update the text to reflect this.

Reply:
That's an omission on my part. I can add that and fix the typo discovered. Thanks!